File: BaekJoon/5052.cpp
#include <iostream>
using namespace std;
struct TrieNode{
    TrieNode* children[10];
    bool isEnd;
    TrieNode() : children(), isEnd(false) {}
    void insert(string &key, int idx){
        if (idx == key.length()) isEnd = true;
        else {
            int next = key[idx] - '0';
            if (children[next] == nullptr) children[next] = new TrieNode;
            children[next]->insert(key, idx + 1);
        }
    }
    bool find(string &key, int idx){
        if (idx == key.length()){
            for(int i = 0; i < 10; i++){
                if (children[i] != nullptr) return true;
            }
            return false;
        } 
        int next = key[idx] - '0';
        return children[next]->find(key, idx + 1);
    }
};
int main(){
    cin.tie(0)->sync_with_stdio(0);
    int T, n; cin >> T;
    while (T--){
        cin >> n;
        TrieNode root;
        string arr[n]; for(string &s : arr){
            cin >> s;
            root.insert(s, 0);
        }
        bool flag = false;
        for(string s : arr) flag = flag || root.find(s, 0);
        cout << (flag ? "NO" : "YES") << "\n";
    }
}

File: BaekJoon/5052.java
import java.io.*;
import java.util.*;

class main {
    public void solution() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        for(int i = 0; i < T; i++){
            int N = Integer.parseInt(br.readLine());
            TrieNode root = new TrieNode();
            ArrayList<String> arr = new ArrayList<>();
            for(int j = 0; j < N; j++){
                arr.add(br.readLine());
                root.insert(arr.get(j), 0);
            }
            boolean flag = false;
            for(String str : arr){
                flag = flag || root.find(str, 0);
            }
            System.out.println((flag ? "NO" : "YES"));
        }
    }

    private class TrieNode {
        private TrieNode[] children;
        private boolean isEnd;

        public TrieNode() {
            children = new TrieNode[10];
            isEnd = false;
        }

        public void insert(String key, int idx){
            if (key.length() == idx) this.isEnd = true;
            else {
                int next = key.charAt(idx) - '0';
                if (children[next] == null) children[next] = new TrieNode();
                children[next].insert(key, idx + 1);
            }
        }

        public boolean find(String key, int idx){
            if (key.length() == idx){
                for(int i = 0; i < 10; i++){
                    if (children[i] != null) return true;
                }
                return false;
            }
            int next = key.charAt(idx) - '0';
            return children[next].find(key, idx + 1);
        }
    }

    public static void main(String[] args) throws Exception{
        new main().solution();
    }
}
